### PR TITLE
Make opers talking through +g controllable by a user mode

### PR DESCRIPTION
**Pull request by spb**

### CODE REVIEW

File: modules/um_callerid.c [hunk 1]
```diff
@@ -63,6 +63,16 @@ um_callerid_modinit(void)
 		return -1;
 	}
 
+	user_modes['@'] = find_umode_slot();
```
Comment (dwfreed) [inline modules/um_callerid.c:66]: Can we... use something other than `@` for this mode? There are 52 alphabetical umodes, surely we haven't used all of them already.

File: modules/um_callerid.c [hunk 2]
```diff
@@ -227,7 +239,34 @@ h_hdl_privmsg_user(void *vdata)
 	data->approved = ERR_TARGUMODEG;
 }
 
+static void
+check_umode_change(void *vdata)
+{
+	hook_data_umode_changed *data = (hook_data_umode_changed *)vdata;
+	bool changed = false;
```
Comment (dwfreed) [inline modules/um_callerid.c:246]: "Variable set but not used" warning from GCC

File: modules/um_callerid.c [hunk 3]
```diff
@@ -227,7 +239,34 @@ h_hdl_privmsg_user(void *vdata)
...
+	if (data->oldumodes & UMODE_OPER && !IsOper(source_p))
+		source_p->umodes &= ~user_modes['@'];
+
+	changed = ((data->oldumodes ^ source_p->umodes) & user_modes['@']);
+
+	if (source_p->umodes & user_modes['@'])
+	{
+		if (!HasPrivilege(source_p, "oper:message"))
+		{
+			sendto_one_notice(source_p, ":*** You need oper:message privilege for +p");
```
Comment (dwfreed) [inline modules/um_callerid.c:261]: Mismatched umode

**Review by dwfreed — CHANGES_REQUESTED** [review]

File: modules/um_callerid.c [hunk 4]
```diff
@@ -227,7 +239,34 @@ h_hdl_privmsg_user(void *vdata)
...
+
+	if (!MyClient(source_p))
+		return;
+
+	if (data->oldumodes & UMODE_OPER && !IsOper(source_p))
+		source_p->umodes &= ~user_modes['@'];
+
+	changed = ((data->oldumodes ^ source_p->umodes) & user_modes['@']);
+
+	if (source_p->umodes & user_modes['@'])
```
Comment (glguy) [inline modules/um_callerid.c:257]: This should be testing for the transition from `-@` to `+@` specifically. Presumably some aspect of `changed` was intended to handle this.

File: modules/um_callerid.c (hunk 1 above)
Comment (spb) [reply modules/um_callerid.c:66]: Yes. I figured there was going to be bikeshedding about the mode character whatever I used, so went with something I know everyone will hate to make sure it happens early.

File: modules/um_callerid.c (hunk 4 above)
Comment (edk0) [reply modules/um_callerid.c]: I think that's wrong; if you lose the oper:message privilege you should get +@ taken away too

File: modules/um_callerid.c [hunk 5]
```diff
@@ -105,7 +117,7 @@ allow_message(struct Client *source_p, struct Client *target_p)
 		return true;
 
 	/* XXX: controversial?  allow opers to send through +g */
-	if (MayHavePrivilege(source_p, "oper:message"))
+	if (IsSetTalkThroughCallerID(source_p))
```
Comment (examknow) [inline modules/um_callerid.c:120]: This makes it impossible for services to speak to a +g user without hacking the services code.

Comment (jesopo) [inline modules/um_callerid.c:120]: does this prevent services messaging a `+g` user?

File: modules/um_callerid.c [hunk 6]
```diff
@@ -227,7 +239,54 @@ h_hdl_privmsg_user(void *vdata)
...
+	if (data->oldumodes & UMODE_OPER && !IsOper(source_p))
+		source_p->umodes &= ~user_modes['M'];
+
+	changed = ((data->oldumodes ^ source_p->umodes) & user_modes['M']);
+
+	if (changed && source_p->umodes & user_modes['M'])
+	{
+		if (!HasPrivilege(source_p, "oper:message"))
+		{
+			sendto_one_notice(source_p, ":*** You need oper:message privilege for +M");
```
Comment (jesopo) [inline modules/um_callerid.c:261]: could we also have this mode timeout like override does?

**Review by jesopo — COMMENTED** [review]
some points from internal discussion

File: modules/um_callerid.c (hunk 6 above)
Comment (glguy) [reply modules/um_callerid.c:261]: I'd rather timeout behaviors be implemented client side if desired. Timeouts at best lead to ugly hacks in bots

**Comment by jesopo** [comment]
is this waiting on anything

**Comment by examknow** [comment]
https://github.com/solanum-ircd/solanum/blob/c77e470ee981bb57fed1af8465e0ed64a6b29b03/modules/um_callerid.c#L259

that should be `MayHavePrivilege`

**Comment by dwfreed** [comment]
> https://github.com/solanum-ircd/solanum/blob/c77e470ee981bb57fed1af8465e0ed64a6b29b03/modules/um_callerid.c#L259
> 
> that should be `MayHavePrivilege`

That only applies to remote clients; the hook tests that the client is local before this.

**Review by dwfreed — APPROVED** [review]

**Pull request merged**